### DISCUSSION TITLE
CommaFeed 5.0 will compile to native code!

### DISCUSSION THREAD

**Discussion by Athou** [Athou/commafeed#1517] [Announcements]
Hey there,

10 years ago, the vast majority of Java applications were deployed on application servers.
When I started the CommaFeed project, I chose the [Dropwizard](https://www.dropwizard.io) framework because it was a breath of fresh air. It was the first framework able to bundle everything needed to create a standalone webapp, without the need for an application server.

10 years later, there are now a lot of frameworks that took inspiration from it. While Dropwizard is still maintained by a couple of very motivated people, it no longer innovates.

I've been prototyping with [Quarkus](https://quarkus.io/) for the last couple of weeks and it's been great. It's a framework supported by RedHat with a very active community that can (among [other things](https://quarkus.io/extensions/)) be compiled to native code thanks to [GraalVM](https://www.graalvm.org/), resulting in **blazing fast startup times** (around 0.3s) and **very low memory footprint** (< 50M).

The [quarkus](https://github.com/Athou/commafeed/tree/quarkus) branch contains the code for the prototype:
- The [README](https://github.com/Athou/commafeed/blob/quarkus/README.md) has been updated with instructions for building and deploying the app
- [CI has been configured](https://github.com/Athou/commafeed/actions?query=branch%3Aquarkus) and builds both native executables (`*-linux` and `*-windows`) and a classic JAR deployment that requires a Java VM (`*-jvm`)
- Docker images are built [here](https://hub.docker.com/r/athou/commafeed/tags?page=&page_size=&ordering=&name=quarkus)

Quarkus does a lot of work at build time to avoid having to do it at runtime. The database choice is made at build time, this is why there are now 4 artifacts generated for the 4 supported databases (`h2` the embedded database, `postgresql`, `mariadb` and `mysql`).

Since [the way the application is configured has changed](https://github.com/Athou/commafeed/blob/quarkus/README.md#configuration), I took the opportunity to revamp the other configuration settings. There were inconsistencies that have been bothering me for a while and now's the chance to cleanup. Some settings have been renamed and they are now nested/grouped by feature.

Overall, I'm very satisfied with this prototype and I'm eager to hear from you. 
There are no database changes between 4.6.0 and this version. Please try it and let me know what you think 😄 

EDIT: CommaFeed 5.0.0 has been released!

**Comment by Athou** [comment]
commafeed.com is now running on Quarkus 🎉 
![image](https://github.com/user-attachments/assets/7f13ad72-990e-4989-a528-0dfa78d966e9)

**Comment by HelgeBoehme** [comment]
Hi, thanks for the update. Do you have plans to support arm64 native images? I saw that the GraalVM supports this platform.
I'm currently using my RPi (clone) as CommaFeed server without docker.

**Reply by Athou** [reply]
GitHub Actions don't have any arm64 runners yet. It's [planned for the end of the year](https://github.blog/news-insights/product-news/arm64-on-github-actions-powering-faster-more-efficient-build-systems/#get-started-using-arm-hosted-runners-today).
I will add support for arm64 native images as soon as it's available :)

**Reply by Athou** [reply]
Looks like arm runners are [available as of today](https://github.blog/changelog/2025-01-16-linux-arm64-hosted-runners-now-available-for-free-in-public-repositories-public-preview/). I'll try to adapt the CI build to also build an arm native docker image.

**Reply by Athou** [reply]
5.4.0 has been released and supports arm64 native images 😊
